### PR TITLE
Add support for tuples and 3-tuples

### PR DESCRIPTION
**Pull request by halfzebra**
Hello friends!

Thanks for this project, looks quite interesting! :raised_hands: 
Here's my first iteration on introducing Tuple support.

Please let me know if this looks alright so far! I'll try to go through the checklist from #34 in this PR.

## Progress

* [x]  add `Tuple (Expr, Expr)` and `Triple (Expr, Expr, Expr)` constructors to `Frontend.Expr` type... solve any compiler errors that arise as a result (`emitExpr` etc)

* [x]  add tests that show all the various ways these can be written in. Don't think about newlines for now (those will be dealt with later), but write tests to see tuples work with and without spaces in between. Eg.:
    
    * [x]  `(1,2)`, `( 1, 2 )`
    * [x]  `(1,2,3)`, `( 1, 2, 3 )`

* [x]  implement type inference - it should be pretty straighforward:
    
    * [x]  if items have types A and B, the tuple has type `(A, B)`. (Yeah, you'll have to create new `Type` constructors to account for this)
    * [x]  similarly for 3-tuples ("triples" - I don't like that name though 🙁 ) - items with types A, B, C ➡️ 3-tuple has type `(A, B, C)`

* [x]  add emit tests too:
    
    * [x]  Elm's `(1, 2)` ➡️ JavaScript's `[1, 2]`, similarly with the 3-tuples

* [x]  implement emitting (`emitExpr`) to pass the emit tests above

* [x]  update the "what's done" table in README for tuples and 3-tuples: parser tests, type inference, emit and emit tests... but not parsers.

### CODE REVIEW

**Comment by Janiczek** [comment]
So far this is great :slightly_smiling_face: :+1:

**Comment by halfzebra** [comment]
Thanks for the feedback, I'm continuing with the implementation! 🤝

**Comment by Janiczek** [comment]
BTW don't be afraid of pushing multiple commits - we have a "squash and merge" strategy on GitHub so when merged it will all be one big commit. Only saying this because force pushes make reviewing a bit difficult :)

File: src/compiler/AST/Common/Type.elm [hunk 1]
```diff
@@ -103,6 +105,29 @@ toString state type_ =
...
 
+        Tuple t1 t2 ->
+            let
+                ( t1String, state1 ) =
+                    toString state t1
+
+                ( t2String, state2 ) =
+                    toString state1 t2
+            in
+            ( "(" ++ t1String ++ "," ++ t2String ++ ")", state2 )
```
Comment (Janiczek) [inline src/compiler/AST/Common/Type.elm]: From `elm repl` it seems like the official implementation adds spaces there to make it look like `( Int, String )`; similarly with the Tuple3

File: src/compiler/AST/Common/Type.elm [hunk 2]
```diff
@@ -103,6 +105,29 @@ toString state type_ =
...
+                ( t1String, state1 ) =
+                    toString state t1
+
+                ( t2String, state2 ) =
+                    toString state1 t2
+
+                ( t3String, state3 ) =
+                    toString state2 t3
+            in
+            ( "(" ++ t1String ++ "," ++ t2String ++ ", " ++ t3String ++ ")", state2 )
```
Comment (Janiczek) [inline src/compiler/AST/Common/Type.elm]: `state3` here

File: src/compiler/Stage/InferTypes/AssignIds.elm [hunk 3]
```diff
@@ -186,3 +186,26 @@ assignIdsWith idSource expr =
...
+        Canonical.Tuple3 e1 e2 e3 ->
+            let
+                (e1_, idSource1) =
+                    assignIdsWith idSource e1
+
+                (e2_, idSource2) =
+                    assignIdsWith idSource1 e2
+
+                (e3_, idSource3) =
+                    assignIdsWith idSource1 e2
```
Comment (Janiczek) [inline src/compiler/Stage/InferTypes/AssignIds.elm]: `idSource2` and `e3` here

**Comment by Janiczek** [comment]
@halfzebra Sorry if I'm flagging stuff that you'd get to eventually. I saw a potentional bug so I made a comment :upside_down_face:

**Comment by halfzebra** [comment]
Thank you very much for the feedback!

I've had to force-push because I've rebased my code on the latest master. Not gonna do it again 🙂

File: src/compiler/Stage/InferTypes/GenerateEquations.elm [hunk 4]
```diff
@@ -249,6 +249,12 @@ generateEquations idSource ( expr, type_ ) =
             , idSource2
             )
 
+        Typed.Tuple _ _ ->
+            ( [], idSource )
+
+        Typed.Tuple3 _ _ _ ->
+            ( [], idSource )
```
Comment (Janiczek) [inline src/compiler/Stage/InferTypes/GenerateEquations.elm]: Those need to return the equations for their inner exprs (`generateEquations idSourceX eX`)

**Comment by halfzebra** [comment]
Hey Martin, thanks for your comment!

I think I've implemented all the things from the checklist, please let me know if there's anything I can add to improve this PR.

**Comment by Janiczek** [comment]
Oh, I just looked at the tests. Obviously it's fine for the parser ones to fail as the parser is not implemented, but look at the Travis log: your tests have `Ok` as part of the test case, while the `master` branch has changed all other tests to `Just`. Result to Maybe. Can you fix that too?

**Comment by halfzebra** [comment]
Good points, I've updated everything and merged master into this branch!

**Comment by Janiczek** [comment]
Merged, thanks again!

**Comment by halfzebra** [comment]
Thanks!

I think I broke the `README.md` a bit, I'll try to fix that asap 😃

**Comment by Janiczek** [comment]
haha I see :joy: OK!

**Comment by Janiczek** [comment]
@halfzebra As I've had a little while during lunch break, I fixed it directly on `master` ;)

**Pull request merged**